From fce275d57fa59b896acd9543d10dc717bdc7f0e0 Mon Sep 17 00:00:00 2001
From: Richard Bradley <richard.bradley@joyent.com>
Date: Fri, 16 Jun 2017 13:46:18 +0000
Subject: [PATCH] MANTA-2038 muppet: haproxies get restarted when new LBs are
 deployed (should only be when new muskies are deployed) MANTA-3240 muppet
 tests are broken Reviewed by: Dave Pacheco <dap@joyent.com> Approved by: Dave
 Pacheco <dap@joyent.com>

---
 lib/watch.js       | 41 +++++++++++++++++++++++++++++++----------
 package.json       |  4 ++++
 test/helper.js     |  1 +
 test/watch.test.js |  6 +++++-
 4 files changed, 41 insertions(+), 11 deletions(-)

diff --git a/lib/watch.js b/lib/watch.js
index f5753ab..0337712 100644
--- a/lib/watch.js
+++ b/lib/watch.js
@@ -14,6 +14,7 @@ var util = require('util');
 var assert = require('assert-plus');
 var once = require('once');
 var vasync = require('vasync');
+var jsprim = require('jsprim');
 
 
 
@@ -120,6 +121,13 @@ Watch.prototype.start = function start(callback) {
                     if (err) {
                         _cb(err);
                     } else if (obj.type !== 'host') {
+                        /*
+                         * webapi and loadbalancer instances both register
+                         * themselves into the same domain, but as different
+                         * types ("host" and "load_balancer", respectively).
+                         * Here we effectively filter out anything but webapi
+                         * instances.
+                         */
                         _cb(null);
                     } else {
                         log.debug({
@@ -143,22 +151,35 @@ Watch.prototype.start = function start(callback) {
                     }, 'watch: getChild failed');
                     self.emit('error', err);
                 } else {
+                    var hosts = [];
                     // This little snippet just drops
                     // nulls and duplicates
-                    self.hosts = [];
                     res.successes.filter(function (h) {
                         return (h);
                     }).forEach(function (h) {
-                        if (self.hosts.indexOf(h) < 0)
-                            self.hosts.push(h);
+                        if (hosts.indexOf(h) < 0)
+                            hosts.push(h);
                     });
-
-                    log.info({
-                        path: self.path,
-                        hosts: self.hosts
-                    }, 'watch: hosts updated');
-
-                    self.emit('hosts', self.hosts);
+                    hosts.sort();
+
+                    /*
+                     * Only emit if the set of webapi instances has
+                     * changed.
+                     */
+                    if (!jsprim.deepEqual(hosts, self.hosts)) {
+                        self.hosts = hosts;
+                        log.info({
+                            path: self.path,
+                            hosts: self.hosts
+                        }, 'watch: hosts updated');
+                        self.emit('hosts', self.hosts);
+                    } else {
+                        log.info({
+                            path: self.path,
+                            new: hosts,
+                            current: self.hosts
+                        }, 'watch: got hosts, but no changes');
+                    }
                 }
             });
         });
diff --git a/package.json b/package.json
index ea974e7..11f1cf5 100644
--- a/package.json
+++ b/package.json
@@ -10,12 +10,16 @@
         "bunyan": "0.22.1",
         "dashdash": "1.4.0",
         "forkexec": "1.1.0",
+        "jsprim": "1.4.0",
         "node-uuid": "1.4.1",
         "once": "1.3.0",
         "vasync": "1.4.3",
         "verror": "1.9.0",
         "zkplus": "0.3.2-smartos-only"
     },
+    "devDependencies": {
+        "nodeunit": "0.7.4"
+    },
     "scripts": {
         "start": "node ./main.js"
     },
diff --git a/test/helper.js b/test/helper.js
index 0c8cd3b..4af483e 100644
--- a/test/helper.js
+++ b/test/helper.js
@@ -51,6 +51,7 @@ function createZkClient(callback) {
                 zk.removeAllListeners('error');
                 callback(null, zk);
         });
+        zk.connect();
 }
 
 
diff --git a/test/watch.test.js b/test/watch.test.js
index a5573ff..e5192bb 100644
--- a/test/watch.test.js
+++ b/test/watch.test.js
@@ -64,7 +64,11 @@ test('create watch', function (t) {
         t.ok(WATCH);
         WATCH.start(function (err) {
                 t.ifError(err);
-                t.end();
+                WATCH.once('hosts', function (hosts) {
+                        t.ok(hosts);
+                        t.equal(hosts.length, 0);
+                        t.end();
+                });
         });
 });
 
-- 
2.21.0

